components/experimental/src/duration/format.rs

```diff
@@ -0,0 +1,528 @@
+            // b. Set signDisplayed to false.
+            *sign_displayed = false;
+        }
+        // 19. Return numericPartsList.
+        Ok(())
+    }
+
+    /// Section 1.1.8
+    /// Returns true if the next smallest unit uses the "fractional" style.
+    fn next_unit_fractional(&self, unit: Unit) -> bool {
+        // 1. Assert: unit is "seconds", "milliseconds", or "microseconds".
+        debug_assert!(matches!(
+            unit,
+            Unit::Second | Unit::Millisecond | Unit::Microsecond
+        ));
+        // 2. If unit is "seconds" and durationFormat.[[MillisecondsStyle]] is "fractional", return true.
+        // 3. Else if unit is "milliseconds" and durationFormat.[[MicrosecondsStyle]] is "fractional", return true.
+        // 4. Else if unit is "microseconds" and durationFormat.[[NanosecondsStyle]] is "fractional", return true.
+        // 5. Return false.
+
+        match unit {
+            Unit::Second if self.fmt.options.second == FieldStyle::Fractional => true,
+            Unit::Millisecond if self.fmt.options.millisecond == FieldStyle::Fractional => true,
+            Unit::Microsecond if self.fmt.options.microsecond == FieldStyle::Fractional => true,
+            _ => false,
+        }
+    }
+
+    fn partition_duration_format_pattern<V: PartsWrite + ?Sized>(
+        &self,
+        sink: &mut V,
+    ) -> core::fmt::Result {
+        // let mut result = Vec::new();
+        // 2. Let signDisplayed be true.
+        let mut sign_displayed = true;
+        // 3. Let numericUnitFound be false.
+        let mut numeric_unit_found = false;
+
+        // 4.  While numericUnitFound is false, repeat for each row in Table 2 in table order, except the header row:
+        for ((unit, style, display), value) in self
+            .fmt
+            .options
+            .iter_units()
+            .into_iter()
+            .zip(self.duration.iter_units())
+        {
+            if numeric_unit_found {
+                break;
+            }
+            // a. Let value be the value of duration's field whose name is the Value Field value of the current row.
+            // b. Let style be the value of durationFormat's internal slot whose name is the Style Slot value of the current row.
+            // c. Let display be the value of durationFormat's internal slot whose name is the Display Slot value of the current row.
+            // d. Let unit be the Unit value of the current row.
+            // e. If style is "numeric" or "2-digit", then
+            if style == FieldStyle::Numeric || style == FieldStyle::TwoDigit {
+                // i. Append FormatNumericUnits(durationFormat, duration, unit, signDisplayed) to result.
+                self.format_numeric_units(unit, &mut sign_displayed, sink)?;
+                // ii. Set numericUnitFound to true.
+                numeric_unit_found = true;
+            }
+            // f. Else,
+            else {
```

Comment:
Nit: The code in the `else` doesn't do anything right now since we are only supporting digital formatting. I think I would rather remove it from the PR. You could have a debug assertion for now if any of the styles don't resolve to Numeric or TwoDigit and have a nonzero value.